### PR TITLE
[UNIT-TESTS] IBC Testing

### PR DESCRIPTION
**Pull request by sampocs**
Closes: #XXX

## What is the purpose of the change
* Many functions in our code base issue IBC or ICA transactions. With the current unit testing framework, issuing these transactions will cause the test to fail. This PR mocks out the necessary Stride state  (e.g. clients/connections/channels) to allow these transactions to succeed.

## Brief Changelog
- Integrated `ibctesting` to our framework through the `AppTestHelper`
- `AppTestHelper` now contains:
  - `StrideChain`: An instance of our chain casted as an ibctesting `TestChain`
  - `HostChain`: A `SimApp` chain casted as an ibctesting `TestChain`
  - `Coordinator`: A broker between chains - responsible for updating clients and keeping chains in sync
- Added helper functions to create transfer or ICA channels (and the clients and connections that go with them)
- Modified `msg_server_claim_undelegated_tokens_test.go` so that it calls the true public facing function (`ClaimUndelegatedTokens`) - which includes an ICA transaction. Previously, the test would call nested functions so it could stop short of the ICA transfer
- Added additional test cases to the claim undelegated tokens test to demonstrate the additional coverage with this new framework

## Breaking Changes
* The following must be changed in existing unit tests
```
s.Ctx → s.Ctx()
s.msgServer → s.GetMsgServer()
```

## Usage
* If you need IBC transfer functionality, call `s.CreateTransferChannel(hostChainId)` in the test's setup function
* If you need ICA transfer functionality, call `s.CreateICAChannel(owner)` in the test's setup function
  * The first time this is called, it will automatically create a transfer channel
* Multiple channels are supported. If needed, you can mock out all 4 ICA channels at once.
* Keep using `s.App` to access stride/stakeibc keeper functions
* Use `s.HostApp` to access the host chains keeper. However, see "Shortcomings/Future Work" below.

## IBC Testing Terminology
* **Coordinator**: Keeps track of two chains
* **Path**: Represents a channel 
* **Endpoint**: Represents an end of a channel and the associated clients and connections

## Testing and Verifying
- Verified by unit tests and integration tests
- Integration test results
```
gaia_tests.bats
 ✓ [INTEGRATION-BASIC] address names are correct
 ✓ [INTEGRATION-BASIC] host zones successfully registered
 ✓ [INTEGRATION-BASIC-GAIA] ibc transfer updates all balances
 ✓ [INTEGRATION-BASIC-GAIA] liquid stake mints stATOM
 ✓ [INTEGRATION-BASIC-GAIA] tokens were transferred to GAIA after liquid staking
 ✓ [INTEGRATION-BASIC-GAIA] tokens on GAIA were staked
 ✓ [INTEGRATION-BASIC-GAIA] redemption works
 ✓ [INTEGRATION-BASIC-GAIA] claimed tokens are properly distributed
 ✓ [INTEGRATION-BASIC-GAIA] rewards are being reinvested, exchange rate updating
 ✓ [NOT-IMPLEMENTED] reinvestment and revenue amounts are correct

10 tests, 0 failures

osmo_tests.bats
 ✓ [INTEGRATION-BASIC] address names are correct
 ✓ [INTEGRATION-BASIC] host zones successfully registered
 ✓ [INTEGRATION-BASIC-OSMO] ibc transfer updates all balances
 ✓ [INTEGRATION-BASIC-OSMO] liquid stake mints stOSMO
 ✓ [INTEGRATION-BASIC-OSMO] tokens were transferred to OSMO after liquid staking
 ✓ [INTEGRATION-BASIC-OSMO] tokens on OSMO were staked
 ✓ [INTEGRATION-BASIC-OSMO] redemption works
 ✓ [INTEGRATION-BASIC-OSMO] claimed tokens are properly distributed
 ✓ [INTEGRATION-BASIC-OSMO] rewards are being reinvested, exchange rate updating
 ✓ [NOT-IMPLEMENTED] reinvestment and revenue amounts are correct

10 tests, 0 failures

juno_tests.bats
 ✓ [INTEGRATION-BASIC] address names are correct
 ✓ [INTEGRATION-BASIC] host zones successfully registered
 ✓ [INTEGRATION-BASIC-JUNO] ibc transfer updates all balances
 ✓ [INTEGRATION-BASIC-JUNO] liquid stake mints stJUNO
 ✓ [INTEGRATION-BASIC-JUNO] tokens were transferred to JUNO after liquid staking
 ✓ [INTEGRATION-BASIC-JUNO] tokens on JUNO were staked
 ✓ [INTEGRATION-BASIC-JUNO] redemption works
 ✓ [INTEGRATION-BASIC-JUNO] claimed tokens are properly distributed
 ✓ [INTEGRATION-BASIC-JUNO] rewards are being reinvested, exchange rate updating
 ✓ [NOT-IMPLEMENTED] reinvestment and revenue amounts are correct

10 tests, 0 failures
```

## Shortcomings/Future Work
* The focus of this PR is on Stride's state. We can extend this in the future to identify changes on the host chain, but that is out of the scope of this PR.
* I have yet to test the IBC transfer functionality with an actual function from our code base. I've only tested with the raw `SendTx` functions. I'll make sure to test it in the next unit test that I write (in a separate repo)

## Documentation and Release Note
  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no) no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no) no
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  / README.md  /   not documented) N/A
  - Does this pull request update existing proto field values (and require a backend and frontend migration)? (yes / no) no
  - Does this pull request change existing proto field names (and require a frontend migration)? (yes / no) no

### CODE REVIEW

**Review by asalzmann — APPROVED** [review]
lgtm! this is going to make it so much easier to write ibc tests, giving us significantly more test coverage 💯

File: app/apptesting/test_helpers.go [hunk 1]
```diff
@@ -1,52 +1,105 @@
...
+	"github.com/Stride-Labs/stride/app"
+)
+
+var (
+	StrideChainID = "STRIDE"
+
+	TestIcaVersion = string(icatypes.ModuleCdc.MustMarshalJSON(&icatypes.Metadata{
+		Version:                icatypes.Version,
+		ControllerConnectionId: ibctesting.FirstConnectionID,
+		HostConnectionId:       ibctesting.FirstConnectionID,
```
[the remark this answers is not in the document]
Comment (sampocs) [reply app/apptesting/test_helpers.go:28]: TLDR, we just need any arbitrary connection ID so I grabbed one that's already a constant in `ibctesting`

The version is only used to confirm the message type (e.g. ica vs ibc transfer). ICA doesn't actually use those fields, it just confirms a valid `icatypes.Metadata` object has been provided, so any connection ID would work here. For context, the "version" for transfers is the string "ics20-1". 

`ibctesting` is the package we're using from `ibc-go`. `FirstConnectionID` is just a constant defined in the package for "connection-0". It wasn't a custom field

Going to leave this as is

File: x/stakeibc/keeper/msg_server_claim_undelegated_tokens_test.go [hunk 2]
```diff
@@ -90,30 +96,37 @@ func (s *KeeperTestSuite) SetupClaimUndelegatedTokens() ClaimUndelegatedTestCase
...
+	s.Require().True(found, "redemption record found")
+	s.Require().False(actualRedemptionRecord.IsClaimable, "redemption record should not be claimable")
+	s.Require().Equal(expectedRedemptionRecord.Amount, actualRedemptionRecord.Amount, "redemption record should not be claimable")
+	// TODO: check callback data here
+}
 
-	redemptionRecord, found := s.App.RecordsKeeper.GetUserRedemptionRecord(s.Ctx, redemptionRecordId)
-	s.Require().True(found)
-	s.Require().False(redemptionRecord.IsClaimable)
+func (s *KeeperTestSuite) TestClaimUndelegatedTokens_SuccessfulTransferMessage() {
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/msg_server_claim_undelegated_tokens_test.go:113]: done

File: x/stakeibc/keeper/msg_server_claim_undelegated_tokens_test.go [hunk 3]
```diff
@@ -25,12 +29,15 @@ type ClaimUndelegatedTestCase struct {
...
+
+	hostChainId := "GAIA"
+	epochNumber := uint64(1)
 	senderAddr := "stride_SENDER"
 	receiverAddr := "cosmos_RECEIVER"
-	redemptionAddr := "cosmos_REDEMPTION"
-	chainId := "GAIA"
-	redemptionRecordId := "GAIA.1.stride_SENDER"
-	connectionId := "connection-0"
+	redemptionAddr := s.IcaAddresses[redemptionIcaOwner]
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/msg_server_claim_undelegated_tokens_test.go:39]: Imo an attribute is a little cleaner than adding a new function. Especially if you wind up using a map under the hood of that function. And I don't think it adds much value to do something like the following for a unit test when the address is fixed.
```
icaAddress, found := s.App.ICAControllerKeeper.GetInterchainAccountAddress(s.Ctx(), ibctesting.FirstConnectionID, portID)
```

Wdyt?

[the remark this answers is not in the document]
Comment (asalzmann) [reply x/stakeibc/keeper/msg_server_claim_undelegated_tokens_test.go:39]: sgtm

**Pull request merged**